### PR TITLE
[Documentation] QAT after PTQ docs update 

### PR DESCRIPTION
**Pull request by daniil-lyakhov**
### Changes

* Documentation structure is updated
* QAT after PTQ documentation is present
### Reason for changes

* To align documentation with the latest changes in torch QAT

### Related tickets

129586

### CODE REVIEW

File: docs/Algorithms.md [hunk 1]
```diff
@@ -1,8 +1,35 @@
...
+  - Symmetric and asymmetric 4 bit compression mode
+  - NF4 compression mode
+  - Mixed precision weights compression
+  - Grouped weights compression
+
+## Training Time Compression
+
+### Quantization
+
+- [Quantization](./usage/training_time_compression/quantization_aware_training/Usage.md) (PyTorch)
```
Comment (MaximProshin) [inline docs/Algorithms.md:21]: Maybe it's better to call it as Quantization Aware Training (QAT).
@alexsu52 what do you think?

File: docs/Algorithms.md [hunk 2]
```diff
@@ -1,8 +1,35 @@
 # Implemented Compression Methods
 
-Each compression method receives its own hyperparameters that are organized as a dictionary and basically stored in a JSON file that is deserialized when the training starts. Compression methods can be applied separately or together producing sparse, quantized, or both sparse and quantized models. For more information about the configuration, refer to the samples.
+## Post-training Compression
 
-- [Quantization](./compression_algorithms/Quantization.md)
+- [Quantization](./usage/post_training_compression/post_training_quantization/Usage.md) (OpenVino, PyTorch, ONNX, TensorFlow)
```
Comment (MaximProshin) [inline docs/Algorithms.md:5]: Maybe it's better to call it as Post Training Quantization (PTQ).
@alexsu52 what do you think?

File: docs/Algorithms.md [hunk 3]
```diff
@@ -1,8 +1,35 @@
...
+  - Symmetric and asymmetric quantization modes
+  - Signed and unsigned
+  - Per tensor/per channel
+  - Exports to OpenVINO format
+
+### Other algorithms
+
+Each compression method in this section receives its own hyperparameters that are organized as a dictionary and basically stored in a JSON file that is deserialized when the training starts. Compression methods can be applied separately or together producing sparse, quantized, or both sparse and quantized models. For more information about the configuration, refer to the samples.
+
+- [Legacy quantization](./usage/training_time_compression/other_algorithms/LegacyQuantization.md) (PyTorch,TensorFlow)
```
Comment (MaximProshin) [inline docs/Algorithms.md:32]: rename it if we decide to do it above

File: docs/Algorithms.md [hunk 4]
```diff
@@ -12,7 +39,7 @@ Each compression method receives its own hyperparameters that are organized as a
   - Automatic bitwidth assignment based on HAWQ
   - Automatic quantization parameter selection and activation quantizer setup based on HW config preset
   - Automatic bitwidth assignment mode AutoQ, based on HAQ, a Deep Reinforcement Learning algorithm to select best mixed precision given quality metric and HW type.
-- [Sparsity](./compression_algorithms/Sparsity.md)
+- [Sparsity](./usage/training_time_compression/other_algorithms/Sparsity.md) (PyTorch,TensorFlow)
```
Comment (MaximProshin) [inline docs/Algorithms.md:42]: maybe better to call it Unstructured Sparsity

File: docs/Algorithms.md [hunk 5]
```diff
@@ -12,7 +39,7 @@ Each compression method receives its own hyperparameters that are organized as a
...
   - Magnitude sparsity
   - Regularization-based (RB) sparsity
-- [Filter pruning](./compression_algorithms/Pruning.md)
+- [Filter pruning](./usage/training_time_compression/other_algorithms/Pruning.md) (PyTorch,TensorFlow)
```
Comment (MaximProshin) [inline docs/Algorithms.md:45]: Filter pruning => Filter Pruning (Structured Sparsity)

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 6]
```diff
@@ -0,0 +1,96 @@
+# Use Neural Network Compression Framework (NNCF) for Quantization Aware Training in PyTorch
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:1]: Use NNCF for Quantization Aware Training in PyTorch

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+
+This is a step-by-step tutorial on how to integrate the NNCF package into the existing PyTorch project.
+The use case implies that the user already has a training pipeline that reproduces training of the model in the floating  point precision and pretrained model.
+The task is to prepare this model for accelerated inference by simulating the compression at train time.
+Please refer to this [document](/docs/usage/training_time_compression/other_algorithms/LegacyQuantization.md) for quantization implementation details.
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:6]: for quantization implementation details => for details of the implementation.

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+model = TorchModel() # instance of torch.nn.Module
+quantized_model = nncf.quantize(model, ...)
+```
+
+### Step 2: Run the training pipeline
+
+At this point, the NNCF is fully integrated into your training pipeline.
+You can run it as usual and monitor your original model's metrics and/or compression algorithm metrics and balance model metrics quality vs. level of compression.
+
+Important points you should consider when training your networks with compression algorithms:
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:24]: Is it valid? Especially sparsity mentions

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+    ```
+
+## Saving and loading compressed models
+
+The complete information about compression is defined by a compressed model and a NNCF config.
+The model characterizes the weights and topology of the network. The NNCF config - how to restore additional modules intoduced by NNCF.
+The NNCF config can be obtained by `quantized_model.nncf.get_config()` on saving and passed to the
+`nncf.torch.load_from_config` helper function to load additional modules from the given NNCF config.
+
+### Saving and loading compressed models in PyTorch
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:51]: I think it's worth to mention why save and load can be needed.

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+quantized_model.load_state_dict(state_dict)
+```
+
+You can save the `compressed_model` object `torch.save` as usual: via `state_dict` and `load_state_dict` methods.
+
+## Advanced usage
+
+### Compression of custom modules
+
+With no target model code modifications, NNCF only supports native PyTorch modules with respect to trainable parameter (weight) compressed, such as `torch.nn.Conv2d`
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:79]: missed dot after the sentence

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+@nncf.register_module(ignored_algorithms=[...])
+class MyModule(torch.nn.Module):
+    def __init__(self, ...):
+        self.weight = torch.nn.Parameter(...)
+    # ...
+```
+
+If registered module should be ignored by specific algorithms use `ignored_algorithms` parameter of decorator.
+
+In the example above, the NNCF-compressed models that contain instances of `MyModule` will have the corresponding modules extended with functionality that will allow NNCF to quantize, sparsify or prune the `weight` parameter of `MyModule` before it takes part in `MyModule`'s `forward` calculation.
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:94]: Is "sparsify or prune" valid here?

**Review by MaximProshin — CHANGES_REQUESTED** [review]

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 11 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md]: No, removed, thanks!

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 9 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md:52]: Done, please check

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 8 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md:24]: Sparsity part is invalid, but dropout disabling is a valid point

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 7 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md]: Done

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 6 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md]: Done

File: docs/Algorithms.md (hunk 5 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: docs/Algorithms.md (hunk 4 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: docs/Algorithms.md (hunk 2 above)
Comment (alexsu52) [reply docs/Algorithms.md:5]: I agree with you.

File: docs/Algorithms.md (hunk 1 above)
Comment (alexsu52) [reply docs/Algorithms.md:21]: I agree with you.

File: README.md [hunk 12]
```diff
@@ -33,17 +33,17 @@ learning frameworks.
```
Comment (alexsu52) [inline README.md:33]: I suggest to split example code in Training-Time Compression section for PyTorch to Quantization-Aware-Training and the config based approach for another scenarios.

File: README.md [hunk 13]
```diff
@@ -33,17 +33,17 @@ learning frameworks.
...
 
 ### Training-Time Compression Algorithms
 
 |Compression algorithm|PyTorch|TensorFlow|
 | :--- | :---: | :---: |
-|[Quantization Aware Training](./docs/compression_algorithms/Quantization.md) | Supported | Supported |
-|[Mixed-Precision Quantization](./docs/compression_algorithms/Quantization.md#mixed-precision-quantization) | Supported | Not supported |
-|[Sparsity](./docs/compression_algorithms/Sparsity.md) | Supported | Supported |
-|[Filter pruning](./docs/compression_algorithms/Pruning.md) | Supported | Supported |
+|Quantization Aware Training | [Supported](./docs/usage/training_time_compression/quantization_aware_training/Usage.md) | [Supported](./docs/usage/training_time_compression/other_algorithms/LegacyQuantization.md) |
```
Comment (alexsu52) [inline README.md:43]: This change is inconsistent. I would suggest to make a start page for Quantization Aware Training and make clickable text "Quantization Aware Training".

File: docs/Algorithms.md (hunk 2 above)
Comment (alexsu52) [inline docs/Algorithms.md:5]: OpenVino -> OpenVINO. Please check the entire document for this typo.

File: docs/Algorithms.md (hunk 1 above)
Comment (alexsu52) [inline docs/Algorithms.md:21]:
```suggestion
- [Quantization Aware Training (QAT)](./usage/training_time_compression/quantization_aware_training/Usage.md) (PyTorch)
```

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 14]
```diff
@@ -0,0 +1,97 @@
...
+
+1. Trace the model via inference in framework operations.
+
+    ```python
+    # To ONNX format
+    import torch
+    torch.onnx.export(quantized_model, dummy_input, './compressed_model.onnx')
+    # To OpenVINO format
+    import openvino as ov
+    ov_quantized_model = ov.convert_model(quantized_model.cpu(), example_input=dummy_input)
```
Comment (alexsu52) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:40]:
```suggestion
    # To OpenVINO format
    import openvino as ov
    ov_quantized_model = ov.convert_model(quantized_model.cpu(), example_input=dummy_input)
```

File: docs/Algorithms.md [hunk 15]
```diff
@@ -1,8 +1,35 @@
...
+  - NF4 compression mode
+  - Mixed precision weights compression
+  - Grouped weights compression
+
+## Training Time Compression
+
+### Quantization
+
+- [Quantization](./usage/training_time_compression/quantization_aware_training/Usage.md) (PyTorch)
+  - No actions required to train a quantized model after the Post-training Quantization
```
[the remark this answers is not in the document]
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: docs/Algorithms.md (hunk 2 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: docs/Algorithms.md (hunk 3 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: README.md (hunk 12 above)
Comment (daniil-lyakhov) [reply README.md:33]: Done, please check

File: README.md (hunk 13 above)
Comment (daniil-lyakhov) [reply README.md]: Done, please check

File: docs/Algorithms.md (hunk 2 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: docs/Algorithms.md (hunk 1 above)
Comment (daniil-lyakhov) [reply docs/Algorithms.md]: Done

File: README.md (hunk 12 above)
Comment (alexsu52) [reply README.md:33]: Looks good for me. @MaximProshin, what do you think?

File: README.md [hunk 16]
```diff
@@ -189,10 +189,46 @@ quantized_model = nncf.quantize(onnx_model, calibration_dataset)
 
 [//]: # (NNCF provides full  [samples]&#40;#post-training-quantization-samples&#41;, which demonstrate Post-Training Quantization usage for PyTorch, TensorFlow, ONNX, OpenVINO.)
 
-### Training-Time Compression
+### Training-Time Quantization
 
 Below is an example of Accuracy Aware Quantization pipeline where model weights and compression parameters may be fine-tuned to achieve a higher accuracy.
 
+<details open><summary><b>PyTorch</b></summary>
```
Comment (alexsu52) [inline README.md:196]:
```suggestion
<details><summary><b>PyTorch</b></summary>
```

File: README.md [hunk 17]
```diff
@@ -189,10 +189,46 @@ quantized_model = nncf.quantize(onnx_model, calibration_dataset)
...
+# Step 2: Initialize NNCF Dataset
+calibration_dataset = nncf.Dataset(dataset_loader, transform_fn)
+# Step 3: Run the quantization pipeline
+quantized_model = nncf.quantize(model, calibration_dataset)
+
+# Now use compressed_model as a usual torch.nn.Module
+# to fine-tune compression parameters along with the model weights
+
+# ... the rest of the usual PyTorch-powered training pipeline
+```
```
Comment (alexsu52) [inline README.md:224]: I suggest to add code with model saving.

File: docs/Algorithms.md [hunk 18]
```diff
@@ -1,8 +1,35 @@
...
+  - Symmetric and asymmetric 4 bit compression mode
+  - NF4 compression mode
+  - Mixed precision weights compression
+  - Grouped weights compression
+
+## Training Time Compression
+
+### Quantization
+
+- [Quantization Aware Training (QAT)](./usage/training_time_compression/quantization_aware_training/Usage.md) (PyTorch)
```
Comment (alexsu52) [inline docs/Algorithms.md:21]:
```suggestion
## Training Time Compression

- [Quantization Aware Training (QAT)](./usage/training_time_compression/quantization_aware_training/Usage.md) (PyTorch)
```

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 19]
```diff
@@ -0,0 +1,94 @@
...
+$$
+\begin{flalign} &
+{input\\_low,input\\_high} = \begin{cases} {input\\_low}',{input\\_high}', \& ZP \in {0,levels-1} \\
+{input\\_low}',{input\\_high}'', \& {input\\_high}'' - {input\\_low}' > {input\\_high}' - {input\\_low}'' \\
+{input\\_low}'',{input\\_high}', \& {input\\_high}'' - {input\\_low}' <= {input\\_high}' - {input\\_low}''
+\end{cases}
+&\end{flalign}
+$$
+
+## Quantization-aware Training Implementation
```
Comment (alexsu52) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:91]: I would suggest to move it on the top of document.

**Comment by MaximProshin** [comment]
@daniil-lyakhov , did you swap Quantization and Usage docs in QAT? Intentionally?

**Comment by daniil-lyakhov** [comment]
> @daniil-lyakhov , did you swap Quantization and Usage docs in QAT? Intentionally?

`Usage.md` is a name for the root document in each directory. `algorithm-name.md` are names for additional details for the `algorithm-name`

File: README.md (hunk 17 above)
Comment (daniil-lyakhov) [reply README.md:241]: Done

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 19 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md]: Done

**Comment by MaximProshin** [comment]
> > @daniil-lyakhov , did you swap Quantization and Usage docs in QAT? Intentionally?
> 
> `Usage.md` is a name for the root document in each directory. `algorithm-name.md` are names for additional details for the `algorithm-name`

Can you check that everything is fine with your recent changes? This is what I see now (previously it was quantization.md):
image

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md [hunk 20]
```diff
@@ -0,0 +1,94 @@
...
+    ```
+
+## Saving and loading compressed models
+
+The complete information about compression is defined by a compressed model and a NNCF config.
+The model characterizes the weights and topology of the network. The NNCF config - how to restore additional modules intoduced by NNCF.
+The NNCF config can be obtained by `quantized_model.nncf.get_config()` on saving and passed to the
+`nncf.torch.load_from_config` helper function to load additional modules from the given NNCF config.
+The quantized model saving allows to load quantized modules to the target model in a new python process and
+requires only example input for the target module, correspondent NNCF config and a quantized model state dict.
```
Comment (MaximProshin) [inline docs/usage/training_time_compression/quantization_aware_training/Usage.md:47]: correspondent NNCF config and a quantized model state dict => corresponding NNCF config and the quantized model state dict

**Review by MaximProshin — APPROVED** [review]
One minor comment above, the rest is ok for me.

File: docs/usage/training_time_compression/quantization_aware_training/Usage.md (hunk 20 above)
Comment (daniil-lyakhov) [reply docs/usage/training_time_compression/quantization_aware_training/Usage.md]: Done
